### ISSUE TITLE
debian: install lsb-release if not installed (requirement)

### ISSUE THREAD

**Issue by nusenu** [nusenu/ansible-relayor#108]
without lsb `ansible_lsb` is not defined and Task
```Ensure torproject.org repository is present (APT)```
will fail

required package on debian/ubuntu: lsb-release

command that should work on the debian/ubuntu target:
```lsb_release -a```

**Comment by ph3x** [comment]
This can be reproduced using ansible 2.2.1.

The VM set up is a debian jessie netinstall minimal with only ssh server selected and
python installed with apt-get later on.

```
me@tor-master:~$ ansible --version
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides


me@at-vie01a-tor01:~$ export LC_ALL=C
me@at-vie01a-tor01:~$ dpkg -l lsb-release
dpkg-query: no packages found matching lsb-release
me@at-vie01a-tor01:~$ lsb_release -a
-bash: lsb_release: command not found
me@at-vie01a-tor01:~$ unset LC_ALL


me@tor-master:~$ ansible -m setup 192.168.1.251 --ask-pass --ask-become-pass
SSH password:
SUDO password[defaults to SSH password]:
192.168.1.251 | SUCCESS => {
    "ansible_facts": {
        "ansible_all_ipv4_addresses": [
            "192.168.1.251"
        ],
        "ansible_all_ipv6_addresses": [
            "fe80::a00:27ff:fe06:ca4"
        ],
        "ansible_architecture": "x86_64",
        "ansible_bios_date": "12/01/2006",
        "ansible_bios_version": "VirtualBox",
        "ansible_cmdline": {
            "BOOT_IMAGE": "/boot/vmlinuz-3.16.0-4-amd64",
            "quiet": true,
            "ro": true,
            "root": "UUID=6567a2a2-dfdd-4e94-9c6e-3d21b873184a"
        },
        "ansible_date_time": {
            "date": "2017-02-16",
            "day": "16",
            "epoch": "1487271228",
            "hour": "19",
            "iso8601": "2017-02-16T18:53:48Z",
            "iso8601_basic": "20170216T195348575687",
            "iso8601_basic_short": "20170216T195348",
            "iso8601_micro": "2017-02-16T18:53:48.575748Z",
            "minute": "53",
            "month": "02",
            "second": "48",
            "time": "19:53:48",
            "tz": "CET",
            "tz_offset": "+0100",
            "weekday": "Donnerstag",
            "weekday_number": "4",
            "weeknumber": "07",
            "year": "2017"
        },
        "ansible_default_ipv4": {
            "address": "192.168.1.251",
            "alias": "eth0",
            "broadcast": "192.168.1.255",
            "gateway": "192.168.1.1",
            "interface": "eth0",
            "macaddress": "08:00:27:06:0c:a4",
            "mtu": 1500,
            "netmask": "255.255.255.0",
            "network": "192.168.1.0",
            "type": "ether"
        },
        "ansible_default_ipv6": {},
        "ansible_devices": {
            "sda": {
                "holders": [],
                "host": "SATA controller: Intel Corporation 82801HM/HEM (ICH8M/ICH8M-E) SATA Controller [AHCI mode] (rev 02)",
                "model": "VBOX HARDDISK",
                "partitions": {
                    "sda1": {
                        "holders": [],
                        "sectors": "7936000",
                        "sectorsize": 512,
                        "size": "3.78 GB",
                        "start": "2048",
                        "uuid": "6567a2a2-dfdd-4e94-9c6e-3d21b873184a"
                    },
                    "sda2": {
                        "holders": [],
                        "sectors": "2",
                        "sectorsize": 512,
                        "size": "1.00 KB",
                        "start": "7940094",
                        "uuid": null
                    },
                    "sda5": {
                        "holders": [],
                        "sectors": "446464",
                        "sectorsize": 512,
                        "size": "218.00 MB",
                        "start": "7940096",
                        "uuid": "859664f8-09aa-4146-8088-c709cc0123cb"
                    }
                },
                "removable": "0",
                "rotational": "1",
                "sas_address": null,
                "sas_device_handle": null,
                "scheduler_mode": "cfq",
                "sectors": "8388608",
                "sectorsize": "512",
                "size": "4.00 GB",
                "support_discard": "0",
                "vendor": "ATA"
            },
            "sr0": {
                "holders": [],
                "host": "IDE interface: Intel Corporation 82371AB/EB/MB PIIX4 IDE (rev 01)",
                "model": "CD-ROM",
                "partitions": {},
                "removable": "1",
                "rotational": "1",
                "sas_address": null,
                "sas_device_handle": null,
                "scheduler_mode": "cfq",
                "sectors": "2097151",
                "sectorsize": "512",
                "size": "1024.00 MB",
                "support_discard": "0",
                "vendor": "VBOX"
            }
        },
        "ansible_distribution": "Debian",
        "ansible_distribution_major_version": "8",
        "ansible_distribution_release": "jessie",
        "ansible_distribution_version": "8.7",
        "ansible_dns": {
            "domain": "ph3x.at",
            "nameservers": [
                "8.8.8.8",
                "8.8.4.4"
            ],
            "search": [
                "ph3x.at"
            ]
        },
        "ansible_domain": "ph3x.at",
        "ansible_env": {
            "HOME": "/home/me",
            "LANG": "de_AT.UTF-8",
            "LANGUAGE": "de_AT:de",
            "LOGNAME": "me",
            "MAIL": "/var/mail/me",
            "PATH": "/usr/local/bin:/usr/bin:/bin:/usr/games",
            "PWD": "/home/me",
            "SHELL": "/bin/bash",
            "SHLVL": "1",
            "SSH_CLIENT": "192.168.1.250 34968 22",
            "SSH_CONNECTION": "192.168.1.250 34968 192.168.1.251 22",
            "SSH_TTY": "/dev/pts/1",
            "TERM": "xterm",
            "USER": "me",
            "_": "/bin/sh"
        },
        "ansible_eth0": {
            "active": true,
            "device": "eth0",
            "features": {},
            "ipv4": {
                "address": "192.168.1.251",
                "broadcast": "192.168.1.255",
                "netmask": "255.255.255.0",
                "network": "192.168.1.0"
            },
            "ipv6": [
                {
                    "address": "fe80::a00:27ff:fe06:ca4",
                    "prefix": "64",
                    "scope": "link"
                }
            ],
            "macaddress": "08:00:27:06:0c:a4",
            "module": "e1000",
            "mtu": 1500,
            "pciid": "0000:00:03.0",
            "promisc": false,
            "speed": 1000,
            "type": "ether"
        },
        "ansible_fips": false,
        "ansible_form_factor": "Other",
        "ansible_fqdn": "at-vie01a-tor01.ph3x.at",
        "ansible_gather_subset": [
            "hardware",
            "network",
            "virtual"
        ],
        "ansible_hostname": "at-vie01a-tor01",
        "ansible_interfaces": [
            "lo",
            "eth0"
        ],
        "ansible_kernel": "3.16.0-4-amd64",
        "ansible_lo": {
            "active": true,
            "device": "lo",
            "features": {},
            "ipv4": {
                "address": "127.0.0.1",
                "broadcast": "host",
                "netmask": "255.0.0.0",
                "network": "127.0.0.0"
            },
            "ipv6": [
                {
                    "address": "::1",
                    "prefix": "128",
                    "scope": "host"
                }
            ],
            "mtu": 65536,
            "promisc": false,
            "type": "loopback"
        },
        "ansible_machine": "x86_64",
        "ansible_machine_id": "fdb9aeedb407494db581bf08c6d13436",
        "ansible_memfree_mb": 1902,
        "ansible_memory_mb": {
            "nocache": {
                "free": 1950,
                "used": 60
            },
            "real": {
                "free": 1902,
                "total": 2010,
                "used": 108
            },
            "swap": {
                "cached": 0,
                "free": 217,
                "total": 217,
                "used": 0
            }
        },
        "ansible_memtotal_mb": 2010,
        "ansible_mounts": [
            {
                "device": "/dev/sda1",
                "fstype": "ext4",
                "mount": "/",
                "options": "rw,relatime,errors=remount-ro,data=ordered",
                "size_available": 2892599296,
                "size_total": 3932315648,
                "uuid": "N/A"
            }
        ],
        "ansible_nodename": "at-vie01a-tor01",
        "ansible_os_family": "Debian",
        "ansible_pkg_mgr": "apt",
        "ansible_processor": [
            "GenuineIntel",
            "Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz",
            "GenuineIntel",
            "Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz",
            "GenuineIntel",
            "Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz",
            "GenuineIntel",
            "Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz"
        ],
        "ansible_processor_cores": 4,
        "ansible_processor_count": 1,
        "ansible_processor_threads_per_core": 1,
        "ansible_processor_vcpus": 4,
        "ansible_product_name": "VirtualBox",
        "ansible_product_serial": "NA",
        "ansible_product_uuid": "NA",
        "ansible_product_version": "1.2",
        "ansible_python": {
            "executable": "/usr/bin/python",
            "has_sslcontext": true,
            "type": "CPython",
            "version": {
                "major": 2,
                "micro": 9,
                "minor": 7,
                "releaselevel": "final",
                "serial": 0
            },
            "version_info": [
                2,
                7,
                9,
                "final",
                0
            ]
        },
        "ansible_python_version": "2.7.9",
        "ansible_selinux": false,
        "ansible_service_mgr": "systemd",
        "ansible_ssh_host_key_dsa_public": "AAAAB3NzaC1kc3MAAACBAJ6t8vBlgYuabsOLQNZwfUadjIQtgLT+HrNpKkbd1uH7tI2LvP0F05/TUa4Rbf3pLv9M3uunqemahRlKIt4uJnEx5fYjq7HQ8cRCMF03dxU+x4NBQOafZJBA7Z0xfPB0Vs8FqskjQk525ZcTC57jh2y+zPnGpN65IgfYZtY8eK/xAAAAFQDtG4jmLZKgSimJmyopsv5srw89IwAAAIB6k5uV202hNqG1klxX2VdCxT/1y0NPEeJGiu7kQ5lqFKDPIqcyZgMaoCaVhNgmVkmmMQ6K90RgKZI5HEfU8qAb7gxz2BOBdrx0MP7L+EgzHrX0vDAqjwDTju1QDI1+B5MG5tyI0tZtZ5ExNql3Up1T4EUC0B7BwZMOxyi1qVA8BAAAAIBaB2I7rr0onATFMdQo8D96QqWyl8lrFE3FpEjZtxZHOiqIe0uVvbJqyjaGGd+oj2kPxGWxeyE1JW4Y9FvfL1CF3KH7pn/WaeoJmOeKsCL524lkqecE1B1lf/mZAVipWwzia+hWdmcy1vnh4a+yomeSXTKLKEUu8bdV5c3XgvdPTw==",
        "ansible_ssh_host_key_ecdsa_public": "AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBETDBPZoAUTWZ0Y7HY4nuOmFkoGlK75lA6srB8ibKx8855HwO31AR49nEHTZnnFDGBVvK8CXkL73qQjM3OgP+H0=",
        "ansible_ssh_host_key_ed25519_public": "AAAAC3NzaC1lZDI1NTE5AAAAIPHW6ZLDUcIu3YuFlz75NYdII3DRSLXQXcJ0+tKKKg4V",
        "ansible_ssh_host_key_rsa_public": "AAAAB3NzaC1yc2EAAAADAQABAAABAQCX/tjpkbhMCuKkd+loUQQsjSxNI3OAdwo/naldNvB1wQ35uUsaT+EckMjlC9g+gdqA6eno8Ih2FGQO9rEUQJe+i+KLrfG4IlS6xF3IIN+dvW6RegwbQuMAiQmFuQ4byxIuHgoe+DOsW4QRWgsR6S4gEqwMEjQwkLzCOkkIuftRyG7Fvivz4+z95lU+uNHX0U2IdKfJUM3LDpQiS7/XzLh+x+nzb/7h/HXNIBDe2KdT5HepuqyIr7XxmrKXog4hEkS3xB8/mCqWZrQGVFE4PLuzyuJNi2alyTSKdXEUjiKZAK6pMN7KVFWHDfvcboEoTHI5rIyXPXfMgA4bh3G0Jmgh",
        "ansible_swapfree_mb": 217,
        "ansible_swaptotal_mb": 217,
        "ansible_system": "Linux",
        "ansible_system_capabilities": [
            ""
        ],
        "ansible_system_capabilities_enforced": "True",
        "ansible_system_vendor": "innotek GmbH",
        "ansible_uptime_seconds": 727,
        "ansible_user_dir": "/home/me",
        "ansible_user_gecos": "me,,,",
        "ansible_user_gid": 1000,
        "ansible_user_id": "me",
        "ansible_user_shell": "/bin/bash",
        "ansible_user_uid": 1000,
        "ansible_userspace_architecture": "x86_64",
        "ansible_userspace_bits": "64",
        "ansible_virtualization_role": "guest",
        "ansible_virtualization_type": "virtualbox",
        "module_setup": true
    },
    "changed": false
}

```

**Comment by ph3x** [comment]
Added run1.txt which fails and run2.txt where it completes successfully then.
[run1.txt](https://github.com/nusenu/ansible-relayor/files/781272/run1.txt)
[run2.txt](https://github.com/nusenu/ansible-relayor/files/781271/run2.txt)

**Comment by nusenu** [comment]
Thanks for the bugreport.

The first time the ansible module `apt` is used it will install its dependencies on the target including `lsb-release` but will not set the `changed` flag to true, so our re-gather handler does not run and the var remains undefined.

I'm not sure if I should force the handler to run everytime (for everyone) or ask first-time users to run twice as a temporary workaround until:
https://github.com/ansible/ansible/issues/21547
gets fixed.

**Comment by ph3x** [comment]
From my point of view I can perfectly live with running the script a second time.
From a general user point of view and ease of use I'd force a run as a workaround and revert it once the bug is fixed.

**Comment by nusenu** [comment]
reference to upstream issue: https://github.com/ansible/ansible/issues/21547

**Comment by nusenu** [comment]
The upstream bug in ansible causing this problem has been fixed and will "probably be in the 2.5.1".
https://github.com/ansible/ansible/issues/21547

Once ansible 2.5.1 is released:
* remove the workaround c457ea9e264d32ed749d527edaa3273e4c9800cc
* increase min. ansible version to 2.5.1

**Comment by nusenu** [comment]
ansible 2.5.3 with the fix has been released, 
test it without c457ea9e264d32ed749d527edaa3273e4c9800cc

if it works, revert c457ea9e264d32ed749d527edaa3273e4c9800cc
and increase min. ansible version to 2.5.3
